source/AI.cpp

```diff
@@ -1366,6 +1548,44 @@ bool AI::FollowOrders(Ship &ship, Command &command) const
 
 
 
+void AI::MoveInFormation(Ship &ship, Command &command)
+{
+	shared_ptr<Ship> parent = ship.GetParent();
+	if(!parent)
+		return;
+	
+	const Body *formationLead = parent.get();
+	const FormationPattern *pattern = ship.GetFormationPattern();
+	
+	// First we retrieve the patterns available for the parent.
+	auto &patterns = formations[formationLead];
+	
+	// Find the existing FormationPositioner for the pattern, or add one if none exists yet.
+	auto insert = patterns.emplace(piecewise_construct,
+		forward_as_tuple(pattern),
+		forward_as_tuple(formationLead, pattern));
```

Comment:
If I'm not missing something: Can this be moved into the formation change function? It seems like we can emplace all of the patterns there already.

This is the only reason this function isn't `const`, so it would be great if this is possible :)

Reply:
Line 1574 has `it->second.NextPosition(&ship)` which determines the position of the ship inside a formation. It  also is non-const.
The main difficulty is that the decission for a ship to fly in formation or to do something else is made inside `AI::MoveEscort()` (and that decission is made every Step()/frame). If a ship is flying in formation, then it needs to get a position in the pattern around the formation leader. If a ship then later decides to do something else (or becomes disabled), then the position in the pattern needs to be freed up for other ships to use.

If we can deduce the ships' decission outside the `AI::MoveEscort()` function, then we can handle the joining and leaving of formations there and handle the formation-positions as `const` inside the Move-functions. Another option would be to also give the formation-position-tracking information as parameter to MoveEsort, but that also doesn't look very nice.

Reply:
I just moved the iterators that calculate the positions to the 'Step()' function. This brings us closer to making the position retrieval a `const` method, since the iterators no longer need to be updated in the `MoveEscorts()`/`MoveInFormation()` functions.